File: heap/703-kth_largest_element_in_a_stream/solution.cpp
class KthLargest {
// Solution explaination:
// Same as the python solution, but wanted to try implementing the same in cpp

public:
    int k;
    priority_queue <int, vector<int>, greater<int>> pq;
    KthLargest(int k, vector<int>& nums) {
        this->k = k;
        for (int i : nums) {
            add(i);
        }
    }
    
    int add(int val) {
        if (this->pq.size() < this->k) {
            this->pq.push(val);
            return this->pq.top();
        }
        if (val > this->pq.top()) {
            this->pq.pop();
            this->pq.push(val);
        }
        return this->pq.top();
    }
};

/**
 * Your KthLargest object will be instantiated and called as such:
 * KthLargest* obj = new KthLargest(k, nums);
 * int param_1 = obj->add(val);
 */

File: heap/703-kth_largest_element_in_a_stream/solution.py
from heapq import heappush, heappop


class KthLargest:
    # Solution explaination

    # Only add to the heap larger element than the current smallest
    # This gurarantees that the smallest element in this heap is the kth largest

    def __init__(self, k: int, nums: List[int]):
        self.numbers = []
        self.k = k
        for num in nums:
            self.add(num)

    def add(self, val: int) -> int:
        if len(self.numbers) < self.k:
            heappush(self.numbers, val)
            return self.numbers[0]

        if val > self.numbers[0]:
            heappop(self.numbers)
            heappush(self.numbers, val)
        return self.numbers[0]


# Your KthLargest object will be instantiated and called as such:
# obj = KthLargest(k, nums)
# param_1 = obj.add(val)
